src/pgduckdb_duckdb.cpp

```diff
@@ -151,9 +161,14 @@ DuckdbCreateConnection(List *rtables, PlannerInfo *planner_info, List *needed_co
 	auto &db = DuckDBManager::Get().GetDatabase();
 	/* Add DuckDB replacement scan to read PG data */
 	auto con = duckdb::make_uniq<duckdb::Connection>(db);
-	con->context->registered_state->Insert(
-	    "postgres_scan", duckdb::make_shared_ptr<PostgresReplacementScanDataClientContextState>(rtables, planner_info,
-	                                                                                            needed_columns, query));
+	auto &context = *con->context;
+	
+	context.registered_state->Insert(
+	    "postgres_state", duckdb::make_shared_ptr<PostgresContextState>(rtables, planner_info, needed_columns, query));
+	auto res = context.Query("set search_path='pgduckdb.main'", false);
```

Comment:
Regarding the tests that are currently failing:
```
2024-09-16 13:40:53.576 UTC [4126] WARNING:  (DuckDB) Catalog Error: Table with name test_table does not exist!
	Did you mean "pg_tables"?
	LINE 1: EXPLAIN SELECT count(*) AS count FROM test_table
```

I guess it's because the search_path in postgres is public. And here we don't include postgres. Given that search_path in postgres can be modified from the default (and often is). I think there are two options:
1. Always send fully-qualified tables to duckdb.
2. Update search_path in duckdb whenever the search_path in postgres changes.

I think 1 is probably the least error, and I think should be fairly easy. I think if we change the search_path of Postgres temporarily to `pg_catalog` while we call `pgduckdb_pg_get_querydef` then all tables will be fully qualified.

Reply:
I thought about this same issue on friday and this morning, I was hoping `pgduckdb_pg_get_querydef` would take care of it but that doesnt seem to be the case.

I thought about solution 2, but I came up with a different solution on duckdb's side which is to add an optional callback to the CatalogSearchPath that is invoked when `CatalogSearchPath::Get()` is called.

We could use that callback to ask postgres what the current `search_path` is and return `CatalogSearchEntry` (catalog + schema) objects for the entries in the `search_path`.

Reply:
> I think 1 is probably the least error, and I think should be fairly easy. I think if we change the search_path of Postgres temporarily to pg_catalog while we call pgduckdb_pg_get_querydef then all tables will be fully qualified.

I don't get what this accomplishes and why that would result in fully qualified tables?

Reply:
@JelteF I'm not sure why this is failing in the CI but it isn't failing locally:
```sql
-- search_path ORDER matters.
SET search_path TO other, public;
SELECT count(*) FROM t;
SELECT count(*) FROM t, public.t;
```

this should highlight the problem but these pass for me locally

Reply:
Ah actually my expected results are wrong, it is broken.
I'm going to see if the solution I proposed above can quickly be implemented to work here

Reply:
> I don't get what this accomplishes and why that would result in fully qualified tables?

The way `pgduckdb_pg_get_querydef` works is that it fully qualifies all tables in the query, unless that is unnecessary because their schema is in the current Postgres search_path. So by temporarily changing the Postgres search_path to `pg_catalog, non of the tables will be in search_path (except for tables in pg_catalog but we need a different solution for that anyway and we dissallow duckdb execution for those tables because of that). To be clear this is because `pg_get_querydef` works that way, and now that we vendored this function in we could change this behaviour, but it seems nicer not to touch the vendored in code unless strictly necessary. Especially because I can get Postgres to create fully qualified names using by applying this tiny diff to the callsite:

```diff
+	auto save_nestlevel = NewGUCNestLevel();
+	SetConfigOption("search_path", "pg_catalog", PGC_USERSET, PGC_S_SESSION);
	const char *query_string = pgduckdb_pg_get_querydef(query, false);
+	AtEOXact_GUC(false, save_nestlevel);
```

For reference the decision on whether to qualify or not qualify can be found here: https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/ruleutils.c#L12996-L12998

Reply:
Ah thanks that makes sense, we only unset the `search_path` for `pg_get_querydef`, not for parsing.